File: stream-server/internal/core/message.go
package core

import "github.com/pion/webrtc/v4"

type Message struct {
	Type           string                     `json:"type"`
	From           string                     `json:"from,omitempty"`
	To             string                     `json:"to,omitempty"`
	Role           string                     `json:"role,omitempty"`
	Name           string                     `json:"name,omitempty"`
	Content        string                     `json:"content,omitempty"`
	State          []RoomState                `json:"state,omitempty"`
	SDP            *webrtc.SessionDescription `json:"sdp,omitempty"`
	ICE            *webrtc.ICECandidateInit   `json:"ice,omitempty"`
	Action         string                     `json:"action,omitempty"`
	IncomingTracks []IncomingTrackMetaData    `json:"incomingTrackMetaData,omitempty"`
	OutgoingTracks []OutgoingTrackMetaData    `json:"outgoingTrackMetaData,omitempty"`
}

type RoomState struct {
	ParticipantID   string `json:"id"`
	ParticipantName string `json:"name"`
	Role            string `json:"role"`
	Status          string `json:"status"`
}

type IncomingTrackMetaData struct {
	ClientTrackID   string `json:"id"`
	ParticipantID   string `json:"participantId"`
	ParticipantName string `json:"participantName"`
	Kind            string `json:"kind"`
}

type OutgoingTrackMetaData struct {
	ClientTrackID   string `json:"clientTrackId"`
	TrackID         string `json:"trackId"`
	ParticipantID   string `json:"participantId"`
	ParticipantName string `json:"participantName"`
	Kind            string `json:"kind"`
}

File: frontend/schemas/message.ts
import { z } from "zod";

export const MessageBase = z.object({
  type: z.string(),
  from: z.string().optional(),
  to: z.string().optional(),
  role: z.string().optional(),
  name: z.string().optional(),
  content: z.string().optional(),
});


export const ParticipantSchema = z.object({
  id: z.string(),
  name: z.string(),
  role: z.string(),
  status: z.string(),
});

export const TrackMetaDataSchema = z.object({
  id: z.string(),
  kind: z.enum(["audio", "video", "screen"]),
  participantId: z.string(),
  participantName: z.string(),
});

export const OutgoingTrackMetaDataSchema = z.object({
  clientTrackId: z.string(),
  trackId: z.string().optional(),
  participantId: z.string(),
  kind: z.enum(["audio", "video", "screen"]),
  participantName: z.string(),
});

export const ChatMessageSchema = MessageBase.extend({
  type: z.literal("chat"),
  content: z.string(),
});
 
export const SDPMessageSchema = MessageBase.extend({
  type: z.literal("sdp"),
  sdp: z.any(), 
  incomingTrackMetaData: z.array(TrackMetaDataSchema).optional(),
  outgoingTrackMetaData: z.array(OutgoingTrackMetaDataSchema).optional(),
});

export const ICEMessageSchema = MessageBase.extend({
  type: z.literal("ice"),
  ice: z.any(), 
});

export const JoinMessageSchema = MessageBase.extend({
  type: z.literal("join"),
});

export const ParicipantLeft = MessageBase.extend({
  type: z.literal("participant_left")
})

export const JoinAckMessageSchema = MessageBase.extend({
  type: z.literal("join_ack"),
  state: z.array(ParticipantSchema),
});

export const ErrorMessageSchema = MessageBase.extend({
  type: z.literal("error"),
  content: z.string(),
});


export const MessageSchema = z.discriminatedUnion("type", [
  ChatMessageSchema,
  SDPMessageSchema,
  ICEMessageSchema,
  JoinMessageSchema,
  JoinAckMessageSchema,
  ErrorMessageSchema,
]);

export type Message = z.infer<typeof MessageSchema>;
export type TrackMetaData = z.infer<typeof TrackMetaDataSchema>;
export type Participant = z.infer<typeof ParticipantSchema>;
export type ChatMessage = z.infer<typeof ChatMessageSchema>;
export type SDPMessage = z.infer<typeof SDPMessageSchema>;
export type ICEMessage = z.infer<typeof ICEMessageSchema>;
export type JoinMessage = z.infer<typeof JoinMessageSchema>;
export type JoinAckMessage = z.infer<typeof JoinAckMessageSchema>;
export type ErrorMessage = z.infer<typeof ErrorMessageSchema>;


